rp2040: Added PIO support

yonsch opened the pull request:
Adds support for PIO for the rp2040 soc. PIO programs are not assembled by Zephyr (currently). They should instead be assembled manually and embedded in source code.
The PIO API should preferably be used to implement drivers. For example, to implement UART over the PIO, it should be implemented as a serial driver, selectable by dts. (this is my suggestion at least).
The PIO API from `pico-sdk` can still be used in application code to implement custom applications.

lairdjm commented on drivers/serial/uart_rpi_pico_pio.c:
```diff
@@ -0,0 +1,172 @@
+	io_rw_8 *uart_rx_fifo_msb;
+	PIO pio = config->pio;
+
+	uart_rx_fifo_msb = (io_rw_8 *)&pio->rxf[RX_STATE_MACHINE] + 3;
```
Comment/details on what's special about the magic number 3?

yonsch replied:
It's a byte pointer to a 32 bit register, so adding 3 gives the MSB of the register, which is denoted in the variable name. But I can document it if it's not clear enough.

rivimey replied:
Just browsing, but yes, the 3 needs documenting. I guess there's no need to be worried about big/little endianness because the PIO is built into the soc, but if it were me I would include a comment including "BYTE_ORDER" just to be clear.

nordicjm reviewed (commented):
Not tested it but looks good!

mbolivar-nordic commented:
@petejohanson would you be able to take a look?

lairdjm commented:
@yonsch please rebase

MrGreensWorkshop commented:
Hi, I wanted to thank you to  @yonsch for this PR, its gives wide perspective about PIO usage on ZephyrOS.
I think it's better to split PIO support and having uart driver as an example in documents. Because you can write almost anything with PIO, but we can't put everything into Zephyr OS right?

Please give some insights. Thank you.

its better to have PIO initiator as driver. (It's good to have consts like frequency in rp2040.dtsi)
```
static inline void __program_init(PIO pio, uint sm, uint offset, uint in_pin, uint in_pin_count, uint out_pin, uint out_pin_count, float frequency) {
  // 1. Define a config object
  pio_sm_config config = __program_get_default_config(offset);  // 2. Set and initialize the input pins
  sm_config_set_in_pins(&config, in_pin);
  pio_sm_set_consecutive_pindirs(pio, sm, in_pin, in_pin_count, 1);
  pio_gpio_init(pio, in_pin);  // 3. Set and initialize the output pins
  sm_config_set_out_pins(&config, out_pin, out_pin_count);
  pio_sm_set_consecutive_pindirs(pio, sm, out_pin, out_pin_count, 0);  // 4. Set clock divider
  if (frequency < 2000) {
    frequency = 2000;
  }
  float clock_divider = (float) clock_get_hz(clk_sys) / frequency * 1000;
  sm_config_set_clkdiv(&config, clock_divider);  // 5. Configure input shift register
  // args: BOOL right_shift, BOOL auto_push, 1..32 push_threshold
  sm_config_set_in_shift(&config, true, false, 32);  // 6. Configure output shift register
  // args: BOOL right_shift, BOOL auto_push, 1..32 push_threshold
  sm_config_set_out_shift(&config, true, false, 32);  // 7. Join the ISR & OSR
  // PIO_FIFO_JOIN_NONE = 0, PIO_FIFO_JOIN_TX = 1, PIO_FIFO_JOIN_RX = 2
  sm_config_set_fifo_join(&config, PIO_FIFO_JOIN_NONE);  // 8. Apply the configuration
  pio_sm_init(pio, sm, offset, &config);  // 9. Activate the State Machine
  pio_sm_set_enabled(pio, sm, true);
}
```
taken from : https://medium.com/geekculture/raspberry-pico-programming-with-pio-state-machines-e4610e6b0f29

yonsch commented:
@MrGreensWorkshop Thanks for your feedback. Do I understand correctly that you are proposing a generic PIO driver instead of a per peripheral one? There were a few reasons for implementing it this way:
1. PIO is most useful for implementing uncommon peripherals such as PS/2 or DVI. Therefore, it was assumed that most users would prefer ready to use peripherals, that could be enabled as if they were "regular" peripherals, instead of having a generic PIO driver which would require some extra work to make into a specific peripheral.
2. Implementing PIO peripherals through specific driver APIs allows users to use those peripherals using existing Zephyr API. For example, now you can use the PIO UART as a console, or write a character with `uart_poll_out`, instead of having to use something like `pio_uart_poll_out` or `pio_write`.
3. We preferred not to bother with distributing `pioasm`, since it was only provided as source code as part of `pico-sdk`. With the current implementation, only the driver's author has to bother with installing the assembler.

And finally, the current implementation still gives access to PIO API, so for special cases a user can still write a custom PIO driver, as part of an out of tree Zephyr driver, or simply as part of the application.

MrGreensWorkshop commented:
@yonsch, Actually I gave my insights from hardware driver point of view. As user point view, probably users prefer to use  ready to use peripherals and I agree with you. Does anyone have any thoughts?

rivimey commented:
I will need to use the PIO driver for my project, in my case because I want a high resolution PWM output (the code for which is one of the RPi samples.)

I had presumed the Z PIO driver would mostly be about managing access to specific PIO instances (reservation), and enabling loading of pioasm programs. Once run, a PIO unit should probably be represented as a distinct device -- perhaps the model to contemplate is that PIO is a "bus" like PCIe via which devices can be accessed, rather than a device itself?

It would be nice if the Z PIO driver made access to all the RPi PIO sample devices as simple as saying "that one", so you could load it up with an API call, but perhaps that's a later step?

MrGreensWorkshop commented:
@yonsch Hi, if it's not too much to ask, Could you provide a sample code how to use this driver.
Thank you.

yonsch commented:
@MrGreensWorkshop Any sample using uart would work, you just need a dts overlay to set up the PIO as a uart. This snippet worked fine when saved as `samples/hello_world/boards/rpi_pico.overlay`:
```
&pinctrl {
	pio_uart0_default: pio_uart0_default {
		group1 {
			pinmux = <PIO0_P3>;
			input-enable;
			bias-pull-up;
		};
		group2 {
			pinmux = <PIO0_P4>;
		};
	};
};
 
&pio0 {
 	status = "okay";
};

&pio_uart0 {
	status = "okay";
	current-speed = <9600>;
	pinctrl-0 = <&pio_uart0_default>;
	pinctrl-names = "default";
};

/ {
	chosen {
		zephyr,console = &pio_uart0;
		zephyr,shell-uart = &pio_uart0;
	};
};

```

dcpleung commented:
This needs to be rebased.

MrGreensWorkshop commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -155,6 +155,40 @@
+			label = "PIO_0";
+			status = "disabled";
+
+			pio_uart0: uart {
```
I think this definition not specificly needs to be here,

MrGreensWorkshop replied:
It's better to have sample code in /samples/boards/

MrGreensWorkshop commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -155,6 +155,40 @@
+			label = "PIO_1";
+			status = "disabled";
+
+			pio_uart1: uart {
```
same as [below](https://github.com/zephyrproject-rtos/zephyr/pull/44316/files#r931179334)

MrGreensWorkshop commented on boards/arm/rpi_pico/doc/index.rst:
```diff
@@ -73,6 +73,9 @@ hardware features:
    * - PWM
      - :kconfig: `CONFIG_PWM`
      - :dtcompatible: `raspberrypi,pico-pwm`
+   * - UART (PIO)
```
I think PIO should be separated from UART. Since PIO is the other peripheral.

MrGreensWorkshop commented on boards/arm/rpi_pico/doc/index.rst:
```diff
@@ -133,3 +136,14 @@ Inside gdb, run:
+
+Zephyr does not (currently) assemble PIO programs. Rather, they should be
+manually assembled and embedded in source code. An example of how this is done
+can be found at `drivers/serial/uart_rpi_pico_pio.c`.
```
Its better to add a explanation about PIO UART driver. UART could be assigned any pin. (as long as 2 pins in a row)

dcpleung commented:
If this is still relevant, please rebase.

dcpleung commented on boards/arm/rpi_pico/doc/index.rst:
```diff
@@ -112,6 +112,17 @@ Default Zephyr Peripheral Mapping:
 - ADC_CH2 : P28
 - ADC_CH3 : P29
 
+Programmable I/O (PIO)
```
Please squash this with the previous commit. This would get rid of that merge artifacts there.

yonsch committed:
drivers: misc: Add driver for RaspberryPi Pico PIO

Added a generic driver for RaspberryPi Pico PIO.
This driver is an intermediate driver for abstracting the PIO
device driver from physical pin configuration.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>
Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>
Signed-off-by: Ionut Catalin Pavel <iocapa@iocapa.com>

yonsch committed:
drivers: serial: Added rpi_pico driver over PIO

Implements a UART driver using PIO. Both PIOs are supported.
Only polling API is supported. Only 8N1 mode is supported.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch committed:
doc: boards: rpi_pico: Added PIO documentation

Adds PIO documentation to the rpi_pico board.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch committed:
samples: boards: rpi_pico: Added uart_pio sample

Added a sample for the rpi_pico board showing how to use the UART
over PIO driver.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch commented:
@soburi I've integrated your suggestion to this PR, with some changes. Most notably, I implemented the pin naming a bit differently (instead of `gpio-names`, a phandle array of nodelabels), to avoid the double definition of the pin number. Please tell me what you think

fabiobaltieri commented on include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h:
```diff
@@ -0,0 +1,87 @@
+	static const uint16_t name ## _program_instructions[] = {	\
+		__VA_ARGS__						\
+	};								\
+	static const struct pio_program name ## _program = {		\
```
Should this be called just `name`? (as in, drop the `_program` suffix) That way it's a bit easier to understand that this is what it taken by `pio_add_program` as an argument just by reading the code.

yonsch replied:
I don't know what it will be useful for, we can always add it later

yonsch replied:
Oh sorry, I though you meant a getter for the instructions themselves. I'll add it

yonsch replied:
done

soburi commented on samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay:
```diff
@@ -0,0 +1,30 @@
+	pinctrl-names = "default";
+	state-machine-num = <2>;
+
+	pins = <&pio1_uart_rx>, <&pio1_uart_tx>;
```
Need to specify which index of the pinmux array.
I tried it and it seems to work with the attached code.

[0001-add-pinmux-cell.patch](https://github.com/zephyrproject-rtos/zephyr/files/10902389/0001-add-pinmux-cell.patch)

yonsch replied:
Why is the index needed?

soburi replied:
The `pinmux` property is an array. There is a possibility to use the second or later elements.

e. g. SPI's pinctrl is usually grouped into OUTPUT and INPUT.
If we implement SPI with PIO, we use all of the OUTPUT group's elements.

https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi#L27-L35

soburi commented on samples/boards/rpi_pico/uart_pio/boards/rpi_pico.overlay:
```diff
@@ -0,0 +1,30 @@
+	pinctrl-names = "default";
+	state-machine-num = <2>;
+
+	pins = <&pio1_uart_rx>, <&pio1_uart_tx>;
```
The right hand side value represent pinmux, not pin.
IMO, The property name should change to `pinmuxs`.

yonsch replied:
done

yonsch commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -193,6 +193,36 @@
+			resets = <&reset RPI_PICO_RESETS_RESET_PIO0>;
+			status = "disabled";
+
+			pio0_uart: uart {
```
agree, done

gmarull commented on dts/bindings/misc/raspberrypi,pico-pio.yaml:
```diff
@@ -0,0 +1,38 @@
+      If the value of this property is greater than the number of elements
+      in the state-machine property, dynamically allocate the missings.
+
+  pinmuxs:
+    type: phandle-array
+    description: |
+      A phandle array of pinmuxes used by the higher level driver. The role of
+      each pinmux is determined by `pinmux-names`. For example, if PIO1 is used
+      as a UART peripheral, `pinmuxs` will be set to `<&pio1_uart_rx>, <&pio1_uart_tx>`.
+
+  pinmux-names:
+    type: string-array
+    description: |
+      The names of the pinmuxes as expected by the higher level driver. For example,
+      if the PIO is used as a UART peripheral, `pinmux-names` will be set to `"rx", "tx"`.
```
-1 on custom pinmux schemes

yonsch replied:
Would the suggestion by @soburi [here](https://github.com/zephyrproject-rtos/zephyr/pull/44316#discussion_r1127166096) solve this?

soburi replied:
@gmarull 
Is this pointing out about that the name `"#pinmuxs-cells"` may use commonly, should not use here?
Does the concern solved if the name change to `pio-pinmuxs`?

gmarull commented on dts/bindings/misc/raspberrypi,pico-pio.yaml:
```diff
@@ -0,0 +1,38 @@
+
+properties:
+  state-machine:
+    type: array
+    description: |
+      Specifies the id of state machines to use.
+      Persistently use the state machine with the specified Id.
```
if this is an array, better use plural name

gmarull commented on drivers/serial/uart_rpi_pico_pio.c:
```diff
@@ -0,0 +1,164 @@
+	const struct pinctrl_dev_config *pcfg;
+};
+
+RPI_PICO_PIO_DECLARE_PROGRAM(uart_tx, 0, 3,
```
it looks like this defines, not declares.

gmarull commented on drivers/serial/uart_rpi_pico_pio.c:
```diff
@@ -0,0 +1,164 @@
+			   offset + RPI_PICO_PIO_GET_WRAP_TARGET(uart_tx),
+			   offset + RPI_PICO_PIO_GET_WRAP(uart_tx));
+
+	pio_sm_set_pins_with_mask(pio, sm, BIT(pin_tx), BIT(pin_tx));
+	pio_sm_set_pindirs_with_mask(pio, sm, BIT(pin_tx), BIT(pin_tx));
```
if this is the only thing required for pins, why can't pinctrl driver do this if function is PIO0/1?

yonsch replied:
There are more than just those two functions that use the pin - e.g. `sm_config_set_out_pins`. It is possible to define the configuration of each state machine in dts and have the generic driver apply that configuration. However, I think this would greatly complicate the dts.

gmarull commented on drivers/serial/uart_rpi_pico_pio.c:
```diff
@@ -0,0 +1,164 @@
+	uint32_t offset;
+	pio_sm_config sm_config;
+
+	offset = pio_add_program(pio, &RPI_PICO_PIO_GET_PROGRAM(uart_tx));
```
suggestion: maybe `GET` macro could actually include `&`

yonsch commented:
Added DNM until we have a verdict from the architecture WG regarding a common API with #53748

soburi reviewed (commented):
@iocapa suggest interesting use case in #56678.

In particular, I think it should consider the possibility of creating multiple instances of PIO program
using multiple state machines.

In this case, it makes sense to specify the SM at the child node.

yonsch commented:
@iocapa Thanks for your effort on #56678 . As this PR is older and has already gone through some reviews, it will be merged first, as the bare minimum required for using PIO. You can then update your PR to add your new features to the existing implementation.
One suggestion that was integrated into this PR was to move the pinctrl to the child nodes rather than the parent PIO device, allowing multiple devices to be defined under one PIO device.
If any of your other changes would require breaking an API later, please say so now, so that I might fix it before this is merged.
@soburi your review would also be appreciated.

soburi commented on drivers/misc/pio_rpi_pico/pio_rpi_pico.c:
```diff
@@ -0,0 +1,55 @@
+
+	retval = pio_claim_unused_sm(config->pio, false);
+	if (retval < 0) {
+		return -EAGAIN;
```
It should better to return -EBUSY.

soburi reviewed (changes requested):
I added one request.
The other parts look good to me.

And I think we need to clarify and respond to the pointing out by @gmarull.

https://github.com/zephyrproject-rtos/zephyr/pull/44316#discussion_r1127455558

soburi commented on drivers/misc/pio_rpi_pico/pio_rpi_pico.c:
```diff
@@ -0,0 +1,55 @@
+
+static int pio_rpi_pico_init(const struct device *dev)
+{
+	/* Nothing to do for now */
```
nitpick:

It seem misleading comment.
It better to remove it.

iocapa commented:
@yonsch 

> EDIT: I see that #56678 does not need to add `pinmux-cells`. However, in that manner the driver forces the pinctrl to be defined in a certain way. In the example above from the UART driver, it is assumed that the pinctrl is called `pio0_uart0_default`, and that the tx and rx pins are at index 0. For peripherals with more pins (e.g. SPI) that would mean that a specific ordering would be needed in the pinmux array. My point is - IMO both implementations use a custom pinctrl scheme

Yes, I don't see a way not to have a custom pinctrl scheme. 
The initial idea in #56678 was to force the user in defining 
the pins in a certain way since there's a 2nd stage pinmux at 
the PIO level, but still reusing existing infrastructure to 
make it look "similar" to other drivers.

`DT_RPI_PICO_PIO_PIN_BY_NAME` allows getting a `pinctrl-N` 
instance by name, indexing into it, getting a specific group 
by name and indexing into it's pinmux. It was made this way 
to not force anything on the user, other than documenting 
the approach taken. 

In the current scheme (`DT_PIO_PIN_BY_NAME`) you are 
pretty much required to name each and every pin. 
This will get pretty ugly for a high bit count parallel port, 
and you will still be constrained by PIO internals on how 
those pins could be assigned.

For example, looks like the current PR needs to define a 4 bit 
parallel port in the following way:
```
pio_bus_default: pio_bus_default {
		pio_bus: group1 {
			#pinmux-cells = <1>;
			pinmux = <PIO1_P2>, <PIO1_P3>, <PIO1_P4>, <PIO1_P5>;
			input-enable;
			bias-pull-up;
		};
	};
...
pinctrl-0 = <&pio_bus_default>;
pinctrl-names = "default";
pinmuxs = <&pio_bus 0>, <&pio_bus 1>, <&pio_bus 2>, <&pio_bus 3>;
pinmux-names = "d0", "d1", "d2", "d3";
```
However, you can't really change the order, since in the driver you will 
need the base, which will always be the lowest index (0, d0, <&pio_bus 0>), 
and any potential order change needs to be made in 3 different places (pinmux, pinmuxs, pinmux-names).
```
uint8_t base = DT_INST_PIO_PIN_BY_NAME(inst, d0);
```

#56678 does the same thing, but directly at the `pinmux` level.
In this case, a potential change only needs to be made at the `pinmux` level.
```
pio_bus_default: pio_bus_default {
		pio_bus {
			pinmux = <PIO1_P2>, <PIO1_P3>, <PIO1_P4>, <PIO1_P5>;
			input-enable;
			bias-pull-up;
		};
	};
...
pinctrl-0 = <&pio_bus_default>;
pinctrl-names = "pinctrl_name";
```

and getting it by:
```
uint8_t base = DT_INST_RPI_PICO_PIO_PIN_BY_NAME(inst, pinctrl_name, pio_bus, 0);
```

Another thing that is unclear to me is how you would use `pinmuxs` and `pinmux-names` 
to implement more than 1 pinctrl handle. Looks like you need to add it to `pinmux-names` 
but with different names for the same pin.
```
pinctrl-0 = <&pio_bus_default>;
pinctrl-1 = <&pio_bus_extra>;
pinctrl-names = "default", "extra";
pinmuxs = <&pio_bus_default_pins 0>, <&pio_bus_default_pins 1>, <&pio_bus_extra_pins 0>, <&pio_bus_extra_pins 1>;
pinmux-names = "d0_default", "d1_default", "d0_extra", "d1_extra";
```

So IMO, a PIO based driver should document the pin allocation 
scheme requirements as there are some constraints, and the scheme should be decided
by the child driver in a way that makes sense for that specific driver.

iocapa commented:
@yonsch 

> If any of your other changes would require breaking an API later, please say so now, so that I might fix it before this is merged.

I don't think you can change the current API to be compatible with #56678, as the changes would break the current implementation.

#56678 cannot really coexist with `PICOSDK_USE_PIO` since you would end up with different paths trying to allocate the same resources. Also i see `PICOSDK_USE_CLAIM` creating unnecessary dependencies since any multicore topics should be solved through zephyr specific APIs.

The PIO parent driver in #56678 was designed just as a resource allocator/interrupt router. All child devices need to use the PIO registers directly.
Technically, the PIO part in picosdk is just a wrapper over basic operations and IMO contains a lot of unnecessary overhead for what is does, also some use-cases will result in runtime issues.

For example, when you allocate SMs to different devices you will need to modify some bits in some registers applicable only to those SMs. This requires atomic operations and `picosdk` functions accessing those registers don't support that. In case of UART, if you have the misfortune to end up in a RMW situation on the same register with another device you could end up in an undefined state.

#56678 uses these to solve that:
```
2.1.2. Atomic Register Access
Each peripheral register block is allocated 4kB of address space, with registers accessed using one of 4 methods,
selected by address decode.
• Addr + 0x0000 : normal read write access
• Addr + 0x1000 : atomic XOR on write
• Addr + 0x2000 : atomic bitmask set on write
• Addr + 0x3000 : atomic bitmask clear on write
```

To summarize, my approach would be to break any dependencies on external rpi code (other that the HW structs). This might prove beneficial in the long run as external SW could change and break existing implementations.

yonsch commented:
@iocapa
> the scheme should be decided by the child driver in a way that makes sense for that specific driver.

Okay, I'm convinced.

> I don't think you can change the current API to be compatible with https://github.com/zephyrproject-rtos/zephyr/pull/56678, as the changes would break the current implementation.

It will break some of the implementation, but the question is if it will break an API, or at least require massive changes. For example, if we merged this PR and then you'll need to move lots of DTS around, that would be counter productive. However if you're just replacing a call to `pio_add_program` with your own implementation, that would be more of an "incremental improvement".
I agree with you that some HAL functionality could be implemented better inside Zephyr, and in the future we might even get rid of the HAL completely (except for hardware definitions as you mentioned).
I would really like to see some of your improvements added, but it is much easier to review in incremental PRs. This PR is pretty old now and has gone through some reviews, so we'd like to merge it. The question is if you see any fundamental problem that will make it too hard for you in the future.

iocapa commented:
> It will break some of the implementation, but the question is if it will break an API, or at least require massive changes. For example, if we merged this PR and then you'll need to move lots of DTS around, that would be counter productive. However if you're just replacing a call to `pio_add_program` with your own implementation, that would be more of an "incremental improvement".

#56678 breaks pretty much everything on the source code side:
[rpi_pico_pio.h](https://github.com/zephyrproject-rtos/zephyr/blob/4fbf811880ab3c5d674cdc0fd1b408dbd673e110/include/zephyr/drivers/misc/rpi_pico_pio/rpi_pico_pio.h)
[rpi_pico_pio_util.h](https://github.com/zephyrproject-rtos/zephyr/blob/4fbf811880ab3c5d674cdc0fd1b408dbd673e110/include/zephyr/drivers/misc/rpi_pico_pio/rpi_pico_pio_util.h)
[rpi_pico_pio.c](https://github.com/zephyrproject-rtos/zephyr/blob/4fbf811880ab3c5d674cdc0fd1b408dbd673e110/drivers/misc/rpi_pico_pio/rpi_pico_pio.c)
[uart_rpi_pico_pio.c](https://github.com/zephyrproject-rtos/zephyr/blob/4fbf811880ab3c5d674cdc0fd1b408dbd673e110/drivers/serial/uart_rpi_pico_pio.c)

So, technically, it's a redesign. On a diff, the only similar thing remaining is `DT_RPI_PICO_PIO_PIN_BY_NAME`.
`pio_rpi_pico_allocate_sm` though it looks similar, is not compatible since there's a use-case where you need to allocate more than 1 state machine in a single call because you need explicit ordering for the `INLINE_OUT_EN / OUT_STICKY` feature.

Another issue is that going with incremental merges will not really work since the UART in #56678 needs the PIO in the same PR. So, it could be an incremental review, but needs to be within a single PR.

As I said in the previous comment, #56678 is not compatible with `pico-sdk`. You could technically have an out-of-tree custom driver and still use the SDK (by manually adjusting the allocation scheme), but you cannot really have a in-tree driver using the SDK.
This break was intentional. While the SDK looks ok and there are some examples, they are not really zephyr driver ready and there are some things that the PIO could do which are not really possible with the SDK.
IMO, adding a new driver should follow the same process as other zephyr drivers do: Read the datasheet / Design / Implement.

So, to summarize:
I'm not against merging the current PR.
Just keep in mind that if #56678 should enter (will be rebased as much as possible), in essence it will be just a big diff on the source side (see the files linked above).

carlescufi commented:
>I'm not against merging the current PR.

Thanks @iocapa. Will merge this one then, and then we can follow-up with improvements.

carlescufi commented:
Removed DNM as the Arch WG [already decided in favor](https://github.com/zephyrproject-rtos/zephyr/pull/44316#issuecomment-1513352591) of this PR

nordicjm commented on boards/arm/rpi_pico/doc/index.rst:
```diff
@@ -109,6 +112,17 @@ Default Zephyr Peripheral Mapping:
 
+Programmable I/O (PIO)
+**********************
+The RP2040 SoC comes with two PIO periherals. These are two simple
```
```suggestion
The RP2040 SoC comes with two PIO peripherals. These are two simple
```

nordicjm reviewed (commented):
Looks good, tested it and it works, with a general serial throughput test the error rate is about 15-20% but there isn't flow control and this is a simple demo so that's acceptable.

gmarull commented on drivers/misc/pio_rpi_pico/pio_rpi_pico.c:
```diff
@@ -0,0 +1,53 @@
+	return config->pio;
+}
+
+static int pio_rpi_pico_init(const struct device *dev)
+{
+	return 0;
+}
```
+1, can be removed

gmarull commented on samples/boards/rpi_pico/uart_pio/prj.conf:
```diff
@@ -0,0 +1 @@
+# nothing here
```
if nothing, just delete the file?

nordicjm replied:
prj.conf files are required even if empty

gmarull commented on samples/boards/rpi_pico/uart_pio/src/main.c:
```diff
@@ -0,0 +1,34 @@
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+
+void main(void)
```
`int`

gmarull commented:
> Removed DNM as the Arch WG [already decided in favor](https://github.com/zephyrproject-rtos/zephyr/pull/44316#issuecomment-1513352591) of this PR

A note here: we should not merge something because it comes first, but evaluate both solutions and merge the one that is better.

MrGreensWorkshop reviewed (approved):
LGTM, all of my concerns were fixed. @yonsch Thank you for your time and effort.

There may be a better solution than this, but in my humble opinion, I think we shouldn't be waiting for those to come for months. (fist PR was made on Mar 30, 2022) Let's give people a chance to use it and give some feedback.

carlescufi commented:
> > Removed DNM as the Arch WG [already decided in favor](https://github.com/zephyrproject-rtos/zephyr/pull/44316#issuecomment-1513352591) of this PR
> 
> A note here: we should not merge something because it comes first, but evaluate both solutions and merge the one that is better.

Agreed, this is not the case here. We took this to the architecture WG, and both PRs were scrutinized. No objections were raised to start with this one and then ask @iocapa and @yonsch to bring features from https://github.com/zephyrproject-rtos/zephyr/pull/56678 in progressively.

gmarull commented:
note: this has been merged with a broken sample https://github.com/zephyrproject-rtos/zephyr/commit/6c93cbf7b496229172fa9508b561417bf91df0ae. Not all comments were addressed, either. @yonsch please submit a follow-up PR or @iocapa feel free to fix in your PR.

yonsch commented:
Well there were still some unresolved issues (like moving the driver under gio/swio/whatever) but I guess @iocapa can take it from here

iocapa commented:
Rebased #56678

Please have a look, and let's continue the discussion there.

Ionut
